client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,165 @@
+
+/**
+ * Module dependencies
+ */
+import config from 'config';
+import localforage from 'localforage';
+import Hashes from 'jshashes';
+
+import debugFactory from 'debug';
+const debug = debugFactory( 'calypso:sync-handler' );
+
+// defaults
+const defaults = config( 'sync-handler-defaults' );
+defaults.driver = localforage[ defaults.driver ];
+
+/**
+ * SyncHandler class
+ */
+export class SyncHandler {
+	/**
+	 * Create a SyncHandler instance
+	 *
+	 * @param {Object} [setup] - sync setup
+	 * @param {Function} handler - wpcom handler
+	 *
+	 * @return {Function} sync-handler wrapper
+	 */
+	constructor( setup, handler ) {
+		if ( 'function' === typeof setup ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.wrapper( handler );
+	}
+
+	wrapper( handler ) {
+		const self = this;
+
+		return function( params, fn ) {
```

Comment:
It may be easier to track down closure down the scope chain if we change `fn` to `wrappedResponseCallback` or some different verbose name